Add /api/sample endpoint, documentation, and CI tests.

elmisback opened the pull request:
This PR adds the first API endpoint (/api/sample) as well as documentation and a basic test for the endpoint. Affected files are summarized below, and a copy of the documentation is shown. @pavpanchekha should probably double-check [my resolution of the conflict in sandbox.rkt](https://github.com/herbie-fp/herbie/compare/main...api-endpoint#diff-091cc6cc2a97075c1438e5b38a160f3eb3bd39d7f75636f3c66da3e6a462b63b), though I think it was trivial.

### affected files
Testing:
[.github/workflows/api-endpoint.yml](https://github.com/herbie-fp/herbie/compare/main...api-endpoint#diff-60b4559165788b2bba9f5821f14834d78ea402ddf084b89517d77c32d7b3957a)
[infra/testApi.mjs](https://github.com/herbie-fp/herbie/compare/main...api-endpoint#diff-1b3d02b00cb2f9f615d671e44cd20a52e89c80ce224280947844ee0147767b2a)

Code:
In src/web/demo.rkt, adds a URL matcher:
`[("api" (string-arg)) #:method (or "post" "get" "put") call-api-fn]`

`call-api-fn` routes requests to an appropriate function. Currently, the only function is `sample`, which runs `get-sample` from src/sandbox.rkt.

Modifies www/doc.html and adds copies of api-endpoint.html to www/doc/1.6 and 1.7, since latest currently points to 1.6.

### documentation
Documentation text is copied below:

## Format for all endpoints
All the endpoints listed below respond to POST requests unless otherwise specified. A typical example of sending a POST request to a running Herbie server is:
![image](https://user-images.githubusercontent.com/8119219/210456080-dcfd74c5-87ce-4cc3-87b5-bb8ce0331d9b.png)

## /api/sample
Call sample on an input like:
![image](https://user-images.githubusercontent.com/8119219/210456133-4491f2e1-7cf1-4bd2-b254-b2692f6450ee.png)

Output from that example, in an unordered array of pairs of the point and exact output calculated:
![image](https://user-images.githubusercontent.com/8119219/210456161-7e1ba490-df37-49b5-a221-1969003dd442.png)

## /api/analyze
Forthcoming.
## /api/alternatives
Forthcoming.
## /api/local-error
Forthcoming.

pavpanchekha commented on .github/workflows/api-endpoint.yml:
```diff
@@ -0,0 +1,54 @@
+        run: racket src/herbie.rkt web --port 8000 &
+      
+      - name: "Pause 4 seconds"
+        run: sleep 4s
```
Kind of ugly. I get why it's there—I'm just trying to think through if this will be flaky. Happy to merge without changing but would appreciate a comment about what our backup is.

elmisback replied:
I agree it's ugly. An alternative is to just wait for an open port with netcat + sleep (and do performance testing elsewhere to catch server hanging on startup). I'll go ahead and change this before merge unless you indicate I should leave it alone. https://stackoverflow.com/questions/27599839/how-to-wait-for-an-open-port-with-netcat

pavpanchekha commented on .github/workflows/api-endpoint.yml:
```diff
@@ -0,0 +1,54 @@
+      #   run: node -e "fetch('https://google.com')"
+
+      - name: "Test the endpoint"
+        run: node infra/testApi.mjs
```
Do you anticipate this file becoming more complex? So far it seems like `curl` would do just as well.

elmisback replied:
I think it makes sense to test endpoint behavior in node rather than with command-line programs, it lets us easily unpack JSON to test whether fields appear correctly.

pavpanchekha replied:
Ok, fine

pavpanchekha commented on src/sandbox.rkt:
```diff
@@ -38,10 +38,34 @@
       ([(pt ex) (in-pcontext context)])
     (values pt ex)))
 
+(define (get-sample test #:seed [seed #f] #:profile [profile? #f])
```
The problem is that we're not checking or handling timeouts or exceptions.We could fix both by leveraging `get-test-result` to do the timeout/exception handling. There's also a potential refactoring to _call_ this from `run-herbie` instead of having duplicate code. If you want I can do these changes.

elmisback replied:
If you want to handle these changes, that would be great!

pavpanchekha replied:
I can do that, ping me when the other stuff is done.

elmisback replied:
@pavpanchekha Sorry this took a while, the other stuff is done!

pavpanchekha commented on src/web/demo.rkt:
```diff
@@ -309,6 +311,35 @@
+    'points result))
+
+(define (call-api-fn req endpoint)
+  (define post-body (request-post-data/raw req))
+  (define post-data (cond (post-body (bytes->jsexpr post-body)) (#t #f)))
+  ;; use hash-ref to access values in the object, e.g.
+  ;; (hash-ref post-data 'key)
+  ;; test via Herbie demo page: (await fetch('/api/sample', {method: 'POST', body: JSON.stringify({a: 42})})).json()
+  (define json-obj (if (equal? endpoint "sample") (handle-sample-endpoint post-data) (hasheq
+                     'endpoint endpoint
+                     'data post-data
+                     )))
+  (response
+    200 #"OK"
+    (current-seconds) APPLICATION/JSON-MIME-TYPE
+    empty
+    (λ (op) (write-json json-obj op)))
```
It looks like this always returns successfully, even if you pass an endpoint that doesn't exist? Why? I think it's probably better to do the routing inside `dispatch`, and if the goal is to abstract over JSON handling write a wrapper function for that.

elmisback replied:
Yeah, that should probably 404 instead of echoing, will fix.

I'll move this stuff into `dispatch` as suggested.

pavpanchekha commented on www/doc/1.6/api-endpoints.html:
```diff
@@ -0,0 +1,59 @@
+<!doctype html>
```
Great idea to add documentation, but you only need to add it to 1.7. Older releases won't have the API!

elmisback replied:
Noted, I'll remove the 1.6 copy.

pavpanchekha commented:
Ok @elmisback I pushed a change to refactor the sandbox. I'm not sure it works right, please test for your use case and review the diff.

If everything passes and the nightlies run go ahead and merge.

pavpanchekha commented:
I guess I broke it. Sorry. Do you think you can debug? It is probably dumb.

elmisback commented:
Will do!

On Mon, Jan 16, 2023, 16:47 Pavel Panchekha ***@***.***>
wrote:

> I guess I broke it. Sorry. Do you think you can debug? It is probably dumb.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/herbie-fp/herbie/pull/533#issuecomment-1384704355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB56HM6WUZH5AAADQYZUBTDWSXT2ZANCNFSM6AAAAAATQKHNJM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

elmisback commented:
I removed the `add-time` call for sampling, since it didn't seem to treat the result of `get-sample` as the right type of object.

I also fixed the endpoint test to avoid hardcoding points, since this would cause failures anytime the sampling method changes.

pavpanchekha commented:
For now removing `add-time` is fine. Longer term we'll probably change `test-result` into some kind of `job-result` struct that all jobs will return (and which will include generic infra like timelines, warnings, and so on). But for now it's fine, and if you think it's ready to merge let's do it.

pavpanchekha commented:
Oh—you probably want to merge in `main` to get the changes to nightlies. That way we can get a nightly pass before merging.

elmisback commented:
Will do! Currently planning to also make a separate branch for the study server so I can keep any hacks isolated.

elmisback commented:
I'll merge after nightlies run.
